### PR TITLE
Features/testblock sync

### PR DESCRIPTION
**Pull request by liuhaox1**

### CODE REVIEW

File: core/block_sync.go [hunk 1]
```diff
@@ -7,7 +7,7 @@
 //
 //   This program is distributed in the hope that it will be useful,
 //   but WITHOUT ANY WARRANTY; without even the implied warranty of
-//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See themarshalTopBlockInfo
```
Comment (bird512) [inline core/block_sync.go]: typo

File: core/block_sync.go [hunk 2]
```diff
@@ -327,16 +328,16 @@ func (bs *blockSyncer) notifyLocalTopBlockRoutine() bool {
...
 }
 
-func (bs *blockSyncer) topBlockInfoNotifyHandler(msg notify.Message) {
+func (bs *blockSyncer) topBlockInfoNotifyHandler(msg notify.Message)error {
 	bnm := notify.AsDefault(msg)
-	if peerManagerImpl.getOrAddPeer(bnm.Source()).isEvil() {
+	if peerManagerImpl.isPeerExists(bnm.Source()) && peerManagerImpl.getOrAddPeer(bnm.Source()).isEvil() {
 		bs.logger.Warnf("block sync this source is is in evil...source is is %v\n", bnm.Source())
-		return
+		return fmt.Errorf("block sync this source is is in evil...source is is %v\n", bnm.Source())
```
Comment (bird512) [inline core/block_sync.go]: should reuse the result of last line

File: core/block_sync.go [hunk 3]
```diff
@@ -327,16 +328,16 @@ func (bs *blockSyncer) notifyLocalTopBlockRoutine() bool {
...
 	}
 	blockHeader, e := bs.unMarshalTopBlockInfo(bnm.Body())
 	if e != nil {
 		bs.logger.Errorf("Discard BlockInfoNotifyMessage because of unmarshal error:%s", e.Error())
-		return
+		return fmt.Errorf("Discard BlockInfoNotifyMessage because of unmarshal error:%s", e.Error())
```
Comment (bird512) [inline core/block_sync.go]: should reuse the result of last line

File: core/block_sync.go [hunk 4]
```diff
@@ -383,8 +385,8 @@ func (bs *blockSyncer) blockResponseMsgHandler(msg notify.Message) {
 
 	blockResponse, e := bs.unMarshalBlockMsgResponse(m.Body())
 	if e != nil {
-		bs.logger.Warnf("Discard block response msg because unMarshalBlockMsgResponse error:%d", e.Error())
-		return
+		bs.logger.Warnf("Discard block response msg because unMarshalBlockMsgResponse error:%s", e.Error())
+		return fmt.Errorf("Discard block response msg because unMarshalBlockMsgResponse error:%s", e.Error())
```
Comment (bird512) [inline core/block_sync.go]: should reuse the result of last line

File: core/block_sync.go [hunk 5]
```diff
@@ -398,8 +400,8 @@ func (bs *blockSyncer) blockResponseMsgHandler(msg notify.Message) {
 
 		// First compare weights
 		if peerTop != nil && localTop.MoreWeight(peerTop.BW) {
-			bs.logger.Debugf("sync block from %v, local top hash %v, height %v, totalQN %v, peerTop hash %v, height %v, totalQN %v", localTop.Hash.Hex(), localTop.Height, localTop.TotalQN, peerTop.BH.Hash.Hex(), peerTop.BH.Height, peerTop.BH.TotalQN)
-			return
+			bs.logger.Debugf("sync block from %v, local top hash %v, height %v, totalQN %v, peerTop hash %v, height %v, totalQN %v", source,localTop.Hash.Hex(), localTop.Height, localTop.TotalQN, peerTop.BH.Hash.Hex(), peerTop.BH.Height, peerTop.BH.TotalQN)
+			return fmt.Errorf("sync block from %v, local top hash %v, height %v, totalQN %v, peerTop hash %v, height %v, totalQN %v", source,localTop.Hash.Hex(), localTop.Height, localTop.TotalQN, peerTop.BH.Hash.Hex(), peerTop.BH.Height, peerTop.BH.TotalQN)
```
Comment (bird512) [inline core/block_sync.go]: should reuse the result of last line

File: core/block_sync_test.go [hunk 6]
```diff
@@ -1,52 +1,441 @@
...
+
+	//only txs
+	blocks = tas_middleware_test.GenOnlyHasTransBlocks()
+	pbblocks = BlockToPbOnlyTxs(blocks)
+	message = tas_middleware_pb.BlockResponseMsg{Blocks: pbblocks}
+	bts,_=proto.Marshal(&message)
+	msg = tas_middleware_test.GenDefaultMessageWithBytes(111,bts)
+	err = blockSyncForTest.blockResponseMsgHandler(msg)
+	if  err == nil{
+		t.Fatalf("except got,but got nil")
```
Comment (bird512) [inline core/block_sync_test.go]: typo

File: core/block_sync_test.go [hunk 7]
```diff
@@ -1,52 +1,441 @@
...
+
+	//tx sign error
+	blocks = tas_middleware_test.GenHashCorrectBlocks()
+	pbblocks = blocksToPb(blocks)
+	message = tas_middleware_pb.BlockResponseMsg{Blocks: pbblocks}
+	bts,_=proto.Marshal(&message)
+	msg = tas_middleware_test.GenDefaultMessageWithBytes(111,bts)
+	err = blockSyncForTest.blockResponseMsgHandler(msg)
+	if  err != nil{
+		t.Fatalf("except err nil,but got error")
```
Comment (bird512) [inline core/block_sync_test.go]: typo

File: core/block_sync_test.go [hunk 8]
```diff
@@ -1,52 +1,441 @@
...
+
+	//correct block hash
+	blocks = tas_middleware_test.GenHashCorrectBlocksByFirstHash(lastBlockHash,200)
+	pbblocks = blocksToPb(blocks)
+	message = tas_middleware_pb.BlockResponseMsg{Blocks: pbblocks}
+	bts,_=proto.Marshal(&message)
+	msg = tas_middleware_test.GenDefaultMessageWithBytes(111,bts)
+	err = blockSyncForTest.blockResponseMsgHandler(msg)
+	if  err != nil{
+		t.Fatalf("except err nil,but got error")
```
Comment (bird512) [inline core/block_sync_test.go]: pypo

File: core/blockchain_test.go [hunk 9]
```diff
@@ -440,7 +440,10 @@ func genTestTx(price uint64, target string, nonce uint64, value uint64) *types.T
 	}
 	tx.Hash = tx.GenHash()
 	sk := common.HexToSecKey(privateKey)
-	sign, _ := sk.Sign(tx.Hash.Bytes())
+	sign, err := sk.Sign(tx.Hash.Bytes())
+	if err !=  nil{
+
```
Comment (bird512) [inline core/blockchain_test.go:444]: need mark error

File: core/fork_processor.go [hunk 10]
```diff
@@ -349,7 +348,7 @@ func (fp *forkProcessor) chainPieceBlockHandler(msg notify.Message) {
 	} else {
 		if len(blocks) == 0 {
 			fp.logger.Errorf("from %v, find ancesotr, but blocks is empty!", source)
-			return
+			return fmt.Errorf("from %v, find ancesotr, but blocks is empty!", source)
```
Comment (bird512) [inline core/fork_processor.go]: should reuse the result of last line

File: core/tx_sync.go [hunk 11]
```diff
@@ -228,11 +229,11 @@ func (ts *txSyncer) getOrAddCandidateKeys(id string) *peerTxsHashes {
 	return v.(*peerTxsHashes)
 }
 
-func (ts *txSyncer) onTxNotify(msg notify.Message) {
+func (ts *txSyncer) onTxNotify(msg notify.Message)error {
 	nm := notify.AsDefault(msg)
 	if peerManagerImpl.getOrAddPeer(nm.Source()).isEvil() {
 		ts.logger.Warnf("tx sync this source is is in evil...source is is %v\n", nm.Source())
-		return
+		return fmt.Errorf("tx sync this source is is in evil...source is is %v\n", nm.Source())
```
Comment (bird512) [inline core/tx_sync.go]: should reuse the result of last line

File: core/tx_sync.go [hunk 12]
```diff
@@ -248,7 +249,7 @@ func (ts *txSyncer) onTxNotify(msg notify.Message) {
 		}
 		if count > txMaxNotifyPerTime {
 			ts.logger.Warnf("Rcv onTxNotify,but count exceeds limit")
-			return
+			return fmt.Errorf("Rcv onTxNotify,but count exceeds limit")
```
Comment (bird512) [inline core/tx_sync.go]: should reuse the result of last line

File: core/tx_sync.go [hunk 13]
```diff
@@ -363,7 +364,7 @@ func (ts *txSyncer) onTxReq(msg notify.Message) {
 		}
 		if count > txPeerMaxLimit {
 			ts.logger.Warnf("Rcv tx req,but count exceeds limit")
-			return
+			return fmt.Errorf("Rcv tx req,but count exceeds limit")
```
Comment (bird512) [inline core/tx_sync.go]: should reuse the result of last line

File: core/tx_sync_test.go [hunk 14]
```diff
@@ -1,62 +1,562 @@
...
 	"testing"
+	"time"
 )
 
-func TestCheckReceivedHashesInHitRate(t *testing.T) {
-	peer := newPeerTxsKeys()
-	txs := makeOrderTransactions()
-	for i := 0; i < len(txs); i++ {
-		peer.addSendHash(txs[i].Hash)
+//func TestCheckReceivedHashesInHitRate(t *testing.T) {
```
Comment (bird512) [inline core/tx_sync_test.go:16]: should remote these unused code

**Review by sesamegu — DISMISSED** [review]
ok

**Review by bird512 — DISMISSED** [review]

**Review by bird512 — APPROVED** [review]

**Pull request merged**